drivers/flash/flash_simulator.c

```diff
@@ -0,0 +1,324 @@
+}
+
+static int flash_sim_erase(struct device *dev, off_t offset, size_t len)
+{
+	u32_t unit_start;
+
+	ARG_UNUSED(dev);
+
+	if (!flash_range_is_valid(dev, offset, len)) {
+		return -EINVAL;
+	}
+
+#ifdef CONFIG_FLASH_SIMULATOR_ERASE_PROTECT
+	if (flash_wp_is_set()) {
+		return -EACCES;
+	}
+#endif
+	/* erase operation must be aligned to the erase unit boundary */
+	if ((offset % CONFIG_FLASH_SIMULATOR_ERASE_UNIT) ||
+	    (len % CONFIG_FLASH_SIMULATOR_ERASE_UNIT)) {
+		return -EINVAL;
+	}
+
+	STATS_INC(flash_sim_stats, flash_erase_calls);
+
+#ifdef CONFIG_FLASH_SIMULATOR_SIMULATE_FAILURES
+	/* generate API failure */
+	int rc = toss(CONFIG_FLASH_SIMULATOR_ERASE_API_FAILURE_RATE_PM);
+
+	if (rc) {
+		STATS_INC(flash_sim_stats, flash_erase_api_errors);
+		return rc;
+	}
+#endif
+
+	unit_start = (offset - CONFIG_FLASH_SIMULATOR_BASE_ADDR) /
+		     CONFIG_FLASH_SIMULATOR_ERASE_UNIT;
+
+	for (u32_t i = 0; i < len / CONFIG_FLASH_SIMULATOR_ERASE_UNIT; i++) {
+
+		STATS_EC_INC(unit_start + i);
+
+#ifdef CONFIG_FLASH_SIMULATOR_SIMULATE_FAILURES
+		if (STATS_EC_GET(unit_start + i) >
+		    CONFIG_FLASH_SIMULATOR_MAX_ERASE_CYCLES) {
+			/* only erase a random number of bytes */
+			memset(FLASH(offset), 0xFF, sys_rand32_get() % len);
+			/* increase the excess erase cycles count */
+			STATS_EEC_INC(unit_start + i);
+			goto wait_and_return;
+		}
+#endif
+	}
+
+	/* erase flash contents by pulling all bits to one */
+	memset(FLASH(offset), 0xFF, len);
+
+#ifdef CONFIG_FLASH_SIMULATOR_SIMULATE_FAILURES
+	int hw_error;
+	u32_t area;
+
+	hw_error = toss(CONFIG_FLASH_SIMULATOR_ERASE_HW_FAILURE_RATE_PM);
+	if (hw_error) {
+		/* pull a few bytes to 0 */
+		area = sys_rand32_get() % len;
+		memset(FLASH(offset + area), 0x00, len - area);
+		STATS_INC(flash_sim_stats, flash_erase_hw_errors);
+	}
+wait_and_return:
+#endif
+
+#ifdef CONFIG_FLASH_SIMULATOR_SIMULATE_TIMING
+	/* wait before returning */
+	k_busy_wait(CONFIG_FLASH_SIMULATOR_MIN_ERASE_TIME_US);
+	STATS_INCN(flash_sim_stats, flash_erase_time_us,
+		   CONFIG_FLASH_SIMULATOR_MIN_ERASE_TIME_US);
+#endif
+
+	return 0;
+}
+
+#ifdef CONFIG_FLASH_PAGE_LAYOUT
+static const struct flash_pages_layout flash_sim_pages_layout = {
+	.pages_count = CONFIG_FLASH_SIMULATOR_FLASH_SIZE,
+	.pages_size = CONFIG_FLASH_SIMULATOR_ERASE_UNIT,
+};
+
+static void flash_sim_page_layout(struct device *dev,
+				  const struct flash_pages_layout **layout,
+				  size_t *layout_size)
+{
+	*layout = &flash_sim_pages_layout;
+	*layout_size = 1;
+}
+#endif
+
+static const struct flash_driver_api flash_sim = {
+	.read = flash_sim_read,
+	.write = flash_sim_write,
+	.erase = flash_sim_erase,
+	.write_protection = flash_wp_set,
+	.write_block_size = CONFIG_FLASH_SIMULATOR_PROG_UNIT,
+#ifdef CONFIG_FLASH_PAGE_LAYOUT
+	.page_layout = flash_sim_page_layout,
+#endif
+};
+
+static int flash_init(struct device *dev)
+{
+	STATS_INIT_AND_REG(flash_sim_stats, STATS_SIZE_32, "flash_sim_stats");
+	memset(mock_flash, 0xFF, ARRAY_SIZE(mock_flash));
```

Comment:
think it is good idea to allow to fill the mock flash by a pattern, like triangle or so, best if filled by noise.

Reply:
That's a good idea. I plan to let the user configure two separate byte patterns:
- one to be used for hardware failures and 
- one to be used when an erase unit has reached its erase cycles limit

Reply:
Oh, I see now you meant to fill the flash with a pattern at initialization..
I think the same could be achieved by writing the flash through the API maybe?